### DISCUSSION TITLE
Can a standalone master server be ran from this?

### DISCUSSION THREAD

**Discussion by filipopo** [OldUnreal/UnrealTournamentPatches#1503] [Q&A]
First of all thanks a lot for your work and apologies if this isn't really the right place to ask but I didn't know where else I could

I'm trying to set up a native linux masterserver for Mobile Forces, an old unreal game (currently we're running everything through wine, as far as I'm aware no linux version of Mobile Forces was released) so I downloaded the 469d patch file from this repo and tried to configure it as a standalone master server like this
```
root@mobilna-sila:/home/mobileforces/server/MobileForces-masterserver/System64# systemctl status mobileforces-masterserver
● mobileforces-masterserver.service - Mobile Forces master server
     Loaded: loaded (/etc/systemd/system/mobileforces-masterserver.service; enabled; vendor preset: enabled)
     Active: active (running) since Wed 2024-01-17 10:29:24 UTC; 3min 12s ago
   Main PID: 2117 (sh)
      Tasks: 2 (limit: 9242)
     Memory: 5.4M
        CPU: 846ms
     CGroup: /system.slice/mobileforces-masterserver.service
             ├─2117 sh /home/mobileforces/server/MobileForces-masterserver/System64/RunMasterServer.sh
             └─2118 ./ucc-bin masterserver

Jan 17 10:29:25 mobilna-sila sh[2118]:    Operation Mode: TCPLink
Jan 17 10:29:25 mobilna-sila sh[2118]: !! Loading TextFile Operation Mode Defaults
Jan 17 10:29:25 mobilna-sila sh[2118]:    TextFile Mode Output:
Jan 17 10:29:25 mobilna-sila sh[2118]: !! Loading TCPLink Operation Mode Defaults
Jan 17 10:29:25 mobilna-sila sh[2118]:    TCPLink Mode Service Port: 28900
Jan 17 10:29:25 mobilna-sila sh[2118]: !! Initializing sockets.
Jan 17 10:29:25 mobilna-sila sh[2118]:    UDP socket bound at port 27900
Jan 17 10:29:25 mobilna-sila sh[2118]:    TCP socket bound at port 28900
Jan 17 10:29:25 mobilna-sila sh[2118]:    Listening on TCP socket.
Jan 17 10:29:25 mobilna-sila sh[2118]: !! Listening for and servicing messages.
```
MasterServer.ini
```
[MasterServer]
GameName=mobileforces
OperationMode=TCPLink
ListenPort=27900

[TCPLink]
TCPPort=28900
```
RunMasterServer.sh
```
#!/bin/sh

command='./ucc-bin masterserver'
until $command; do
  echo "Server crashed with exit code $?, restarting!" >&2
  sleep 1
done
```
ucc seems to be working and says that it's listening for connections but from the game client I get messages about the connection timing out to the master server

Now I'm not sure if this is normal or not but when I try to test the master server outside of the game client with this command `telnet mf.nofisto.com 28900`, the connection is established but I get no response:

<img width="399" alt="Screenshot 2024-01-17 at 11 48 19" src="https://github.com/OldUnreal/UnrealTournamentPatches/assets/4060824/0b01632b-d0c8-4a31-a1b8-ee40153ea0b8">

I was expecting to see the `\basic\\secure\wookie` security challenge and I thought that all UT99 games send this (source: [UT99Tools](https://www.powershellgallery.com/packages/Ut99Tools/0.1.1/Content/Ut99Tools.psm1))

For reference this is what it should be like when it's working:
![Screenshot_2023-12-06_at_02 52 47](https://github.com/OldUnreal/UnrealTournamentPatches/assets/4060824/7b925df3-aa2c-4693-9658-b96b665393cb)

Any help in setting this up would be much appreciated!

p.s If anyone is interested, I've set up this master server alongside the game server and an Apache2 web server (FastDL and game management panel) to be automatically provisioned through Ansible: [MF-ansible](https://github.com/filipopo/MF-ansible)

**Comment by SeriousBuggie** [comment]
On windows it work fine. Can be some firewall issue?

**Reply by filipopo** [reply]
Hey @SeriousBuggie Thanks for the reply, I've already hosted a master server on the same server so I didn't think this was the case but I tested it:
I downloaded OldUnreal-UTPatch469d-Windows.zip and modified RunMasterServer.sh to run with wine
```
#!/bin/sh

command='wine UCC.exe masterserver'
until $command; do
  echo "Server crashed with exit code $?, restarting!" >&2
  sleep 1
done
```
The windows version successfully registered my game server:
```
=======================================
ucc.exe: UnrealOS execution environment
Copyright 1999 Epic Games Inc
=======================================

Executing Class IpDrv.MasterServerCommandlet
UMasterServerCommandlet: Unreal Engine Master Server
Copyright 1999 Epic Games, Inc.

!! Loading Config: MasterServer.ini
   GameName: mobileforces
   Operation Mode: TCPLink
!! Loading TextFile Operation Mode Defaults
   TextFile Mode Output:
!! Loading TCPLink Operation Mode Defaults
   TCPLink Mode Service Port: 28900
!! Initializing sockets.
WinSock: version 1.1 (2.2), MaxSocks=32767, MaxUdp=65467
   UDP socket bound at port 27900
   TCP socket bound at port 28900
   Listening on TCP socket.
!! Listening for and servicing messages.
!! Creating mastermap for version 0.
Appr:1 Pend:0 Ign:0 Rej:0 BpsIn:3 BpsOut:0 Req/Sec:0
```
<img width="639" alt="Screenshot 2024-01-18 at 22 14 54" src="https://github.com/OldUnreal/UnrealTournamentPatches/assets/4060824/a5b677ea-f5cf-4991-92e3-6280220b0ccb">

By comparison this is the output from the Linux versions (I tested both the amd64 and x86 versions):
```
WARNING: using preference directory: /root/.utpg/System/
WARNING: Using system ini file: /root/.utpg/System/UnrealTournament.ini
WARNING: Using user ini file: /root/.utpg/System/User.ini
=======================================
ucc.exe: UnrealOS execution environment
Copyright 1999 Epic Games Inc
=======================================

Executing Class IpDrv.MasterServerCommandlet
UMasterServerCommandlet: Unreal Engine Master Server
Copyright 1999 Epic Games, Inc.

!! Loading Config: MasterServer.ini
   GameName: mobileforces
   Operation Mode: TCPLink
!! Loading TextFile Operation Mode Defaults
   TextFile Mode Output: 
!! Loading TCPLink Operation Mode Defaults
   TCPLink Mode Service Port: 28900
!! Initializing sockets.
   UDP socket bound at port 27900
   TCP socket bound at port 28900
   Listening on TCP socket.
!! Listening for and servicing messages.
Appr:0 Pend:0 Ign:0 Rej:0 BpsIn:0 BpsOut:0 Req/Sec:0      
```
<img width="399" alt="Screenshot 2024-01-17 at 11 48 19" src="https://github.com/OldUnreal/UnrealTournamentPatches/assets/4060824/0b01632b-d0c8-4a31-a1b8-ee40153ea0b8">

Could you please verify if the Linux version works fine?, also I don't know does it matter but this message doesn't appear in the Linux version: `!! Creating mastermap for version 0.`

And btw this is my server info
```
root@mobilna-sila:/home/mobileforces/test/OldUnreal-Windows/System# hostnamectl
 Static hostname: mobilna-sila
       Icon name: computer-vm
         Chassis: vm
      Machine ID: df5a661c34ee440796ffd8a330b9de74
         Boot ID: d528a718a6644c8c9fde4d6fd3d8a7f1
  Virtualization: kvm
Operating System: Ubuntu 22.04.3 LTS              
          Kernel: Linux 5.15.0-91-generic
    Architecture: x86-64
 Hardware Vendor: Hetzner
  Hardware Model: vServer
```

**Reply by filipopo** [reply]
A wild guess but... is the cipher different on Windows and Linux?

**Reply by filipopo** [reply]
Bump, just tried it with 469e rc7 and I'm still having the same issue with the Linux ucc but the Windows UCC.exe works fine
